File: Java/Sorting/radix_sort.java
import java.util.Arrays;
import java.util.Scanner;

public class radix_sort {
    public static void main(String[] args) {
        // import scanner to take user input
        Scanner sc = new Scanner(System.in);
        System.out.println("enter the size of the array");
        int size = sc.nextInt();
        System.out.println("enter the element in the array");
        int[] arr = new int[size];
        // store the inputed elements in the array
        for(int i = 0; i<size; i++) {
            arr[i] = sc.nextInt();
        }
        // call our sorting function
        radixSort(arr);
        System.out.println(Arrays.toString(arr));
    }
    public static void radixSort(int[] arr) {
        // search for the maximum number to know the number of digits
        int max = getMax(arr);
        // do counting sort here
        for (int exp = 1; max/exp > 0; exp *= 10) {
            countingSort(arr, exp);
        }
    }
    
    public static void countingSort(int[] arr, int exp) { // sort the array by each digit starting with the least significant digit in the array
        int[] output = new int[arr.length]; // the final array in which we will store our answer
        int[] count = new int[10]; // the counting array

        // store all the counting occourences in count array
        for (int i = 0; i < arr.length; i++) {
            count[(arr[i]/exp)%10]++;
        }

        // change our element at count[i] to count[i-1] so that this contains actual position of the digit in output array
        for (int i = 1; i < 10; i++) {
            count[i] += count[i - 1];
        }

        // create our output array
        for (int i = arr.length - 1; i >= 0; i--) {
            output[count[(arr[i]/exp)%10] - 1] = arr[i];
            count[(arr[i]/exp)%10]--;
        }

        // store the output array in our ans array, so that numbers that are sorted now can be stored
        for (int i = 0; i < arr.length; i++) {
            arr[i] = output[i];
        }
    }

    // normal function to find the max value in our array
    public static int getMax(int[] arr) {
        int max = arr[0];
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] > max) {
                max = arr[i];
            }
        }
        return max;
    }
}

File: C/Sorting/radix_sort.c
// Radix sort is a non-comparative sorting algorithm that sorts data by grouping individual digits or characters together according to their place value. It is a linear-time sorting algorithm that has a complexity of O(kn), where k is the maximum number of digits or characters in the input.

// Best Time Complexity: O(d(n+k)) 
// Average Time Complexity: O(d(n+k))
// Worst Time Complexity: O(d(n+k))
// here d is the maximum number of digits in the input numbers, n is the number of elements in the array, and k is the range of the input numbers.

#include <stdio.h>
#include <stdlib.h>

// A utility function to get maximum value in arr[]
int getMax(int arr[], int n) {
    int max = arr[0];
    for (int i = 1; i < n; i++) {
        if (arr[i] > max) {
            max = arr[i];
        }
    }
    return max;
}

// A function to do counting sort of arr[] according to
// the digit represented by exp.
void countSort(int arr[], int n, int exp) {
    int output[n]; // output array
    int i, count[10] = {0};

    // Store count of occurrences in count[]
    for (i = 0; i < n; i++) {
        count[(arr[i] / exp) % 10]++;
    }

    // Change count[i] so that count[i] now contains actual
    // position of this digit in output[]
    for (i = 1; i < 10; i++) {
        count[i] += count[i - 1];
    }

    // Build the output array
    for (i = n - 1; i >= 0; i--) {
        output[count[(arr[i] / exp) % 10] - 1] = arr[i];
        count[(arr[i] / exp) % 10]--;
    }

    // Copy the output array to arr[], so that arr[] now
    // contains sorted numbers according to current digit
    for (i = 0; i < n; i++) {
        arr[i] = output[i];
    }
}

// Radix sort
void radixSort(int arr[], int n) {
    // Find the maximum number to know number of digits
    int m = getMax(arr, n);

    // Do counting sort for every digit
    for (int exp = 1; m / exp > 0; exp *= 10) {
        countSort(arr, n, exp);
    }
}

int main() {
    int n;
    printf("Enter the number of elements: ");
    scanf("%d", &n);

    int arr[n];
    printf("Enter %d elements: ", n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }

    radixSort(arr, n);

    printf("Sorted array: ");
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    return 0;
}

